hazelcast/src/test/java/com/hazelcast/collection/impl/AbstractCollectionBackupTest.java

```diff
@@ -48,15 +44,77 @@
     protected Config config = new Config();
 
-    protected final void testBackups(CollectionType collectionType, String collectionName, int partitionId) {
+    /**
+     * Returns an instance of a Hazelcast collection, e.g. {@link com.hazelcast.core.ISet}.
+     *
+     * @param instance the {@link HazelcastInstance} to retrieve the collection from
+     * @param name     the name of the collection
+     * @return an instance of a Hazelcast collection
+     */
+    protected abstract Collection<Integer> getHazelcastCollection(HazelcastInstance instance, String name);
+
+    /**
+     * Returns the internal items of a Hazelcast collection, e.g. {@link CollectionItem}.
+     *
+     * @param instance the {@link HazelcastInstance} to retrieve the internal items from
+     * @param name     the name of the collection
+     * @return the internal items of a Hazelcast collection
+     */
+    protected abstract Collection<?> getCollectionItems(HazelcastInstance instance, String name);
+
+    /**
+     * Returns the backup values of a Hazelcast collection.
+     *
+     * @param instance the {@link HazelcastInstance} to retrieve the backup values from
+     * @param name     the name of the collection
+     * @return the backup values of a Hazelcast collection
+     */
+    protected abstract Collection<Integer> getBackupCollection(HazelcastInstance instance, String name);
+
+    protected final void testBackupPromotion(String collectionName, int partitionId) {
+        int expectedCount = ITEM_COUNT * 2;
+
+        HazelcastInstance[] instances = factory.getAllHazelcastInstances().toArray(new HazelcastInstance[0]);
+        HazelcastInstance ownerInstance = getBackupInstance(instances, partitionId, 0);
+        HazelcastInstance backupInstance = getBackupInstance(instances, partitionId, 1);
+
+        // we terminate the ownerInstance, so the backups on backupInstance have to be promoted
+        ownerInstance.getLifecycleService().terminate();
+        instances = factory.getAllHazelcastInstances().toArray(new HazelcastInstance[0]);
+        waitAllForSafeState(instances);
+
+        // we add additional items to the collection, to ensure that the itemId has been set properly during the promotion
+        Collection<Integer> collection = getHazelcastCollection(backupInstance, collectionName);
+        for (int i = ITEM_COUNT; i < expectedCount; i++) {
+            collection.add(i);
+        }
+        assertEqualsStringFormat("collection should contain %d items, but found %d", expectedCount, collection.size());
+
+        // we check the size and itemIDs of the internal collection
+        Set<Long> itemIds = new HashSet<Long>(expectedCount);
+        Collection<?> items = getCollectionItems(backupInstance, collectionName);
+        assertEqualsStringFormat("internal collection should contain %d items, but found %d", expectedCount, items.size());
+        for (Object item : items) {
+            long itemId = INVALID_ITEM_ID;
```

Comment:
this looks fragile. the test is making quite strong assumptions about the implementation.
any idea how to improve it?